mark dirty section and rows cleaned after they are rendered

heywarddai opened the pull request:
Currently, if dataSource is changed, the dirty sections and rows will never get "cleaned", causing them to be re-rendered every frame afterwards.
After a row or section is rendered, we "clean" the corresponding item in the dataSource's _dirtySections and _dirtyRows.

Test plan:
Log to console each time a row renders
Update datasource for a single row
Check that the row only ever re-renders once

This is related to pr: #7405
but the fix is moved out of rowShouldUpdate in ListViewDataSource.

vevo-heyward-dai commented:
related to this issue: https://github.com/facebook/react-native/issues/5255

ericvicenti commented on Libraries/CustomComponents/ListView/ListViewDataSource.js:
```diff
@@ -319,6 +319,27 @@ class ListViewDataSource {
+    */
+  dirtyRowCleaned(sectionIndex: number, rowIndex: number): void {
+    if (this._dirtyRows.length > sectionIndex) {
+      var section = this._dirtyRows[sectionIndex];
```
should be `const` instead of `var`

heywarddai commented:
@ericvicenti 
I originally tried doing this in componentDidMount but couldn't figure out how to only "clean" the sections/rows that were actually rendered and keep the un-rendered ones still dirty. I guess we could keep an set of "to-be-cleaned" section/rows, update that in the render() method and then clean those in componentDidMount()..?

updated pr to fix:
-updated the `var` section declaration to `const`
-moved the dirty section cleaning code to be inside the `this.props.renderSectionHeader` check. In the original code, the list view would not have rendered the section if a renderSectionHeader method was provided to it later.

vevo-heyward-dai commented:
@ericvicenti not sure why tests are failing after merging in master. seems like some android test device is down?
